util/perl/OpenSSL/paramnames.pm

```diff
@@ -0,0 +1,663 @@
+    'LIBSSL_RECORD_LAYER_PARAM_MAX_FRAG_LEN' =>   "max_frag_len",
+    'LIBSSL_RECORD_LAYER_PARAM_MAX_EARLY_DATA' => "max_early_data",
+    'LIBSSL_RECORD_LAYER_PARAM_BLOCK_PADDING' =>  "block_padding",
+);
+
+# Generate string based macros for public consumption
+sub generate_public_macros {
+    my @macros = ();
+
+    foreach my $name (keys %params) {
+        my $val = $params{$name};
+        my $def = '# define OSSL_' . $name . ' ';
+
+        if (substr($val, 0, 1) eq '*') {
+            $def .= 'OSSL_' . substr($val, 1);
+        } else {
+            $def .= '"' . $val . '"';
+        }
+        push(@macros, $def)
+    }
+    return join("\n", sort @macros);
+}
+
+# Generate number based macros for internal use
+# The numbers are unique per string
+sub generate_internal_macros {
+    my @macros = ();
+    my $count = 0;
+    my %reverse;
+
+    # Determine the number for each unique string
+    # Sort the names to improve the change of cache coherency
+    foreach my $name (sort keys %params) {
+        my $val = $params{$name};
+
+        if (substr($val, 0, 1) ne '*' and not defined $reverse{$val}) {
+            $reverse{$val} = $count++;
+        }
+    }
+
+    # Output the defines
+    foreach my $name (keys %params) {
+        my $val = $params{$name};
+        my $def = '#define PIDX_' . $name . ' ';
+
+        if (substr($val, 0, 1) eq '*') {
+            $def .= 'PIDX_' . substr($val, 1);
+        } else {
+            $def .= $reverse{$val};
+        }
+        push(@macros, $def)
+    }
+    return "#define NUM_PIDX $count\n\n" . join("\n", sort @macros);
+}
+
+sub generate_trie {
+    my %trie;
+    my $nodes = 0;
+    my $chars = 0;
+
+    foreach my $name (keys %params) {
+        my $val = $params{$name};
+        if (substr($val, 0, 1) ne '*') {
+            my $cursor = \%trie;
+
+            $chars += length($val);
+            for my $i (0 .. length($val) - 1) {
+                my $c = substr($val, $i, 1);
+
+                if (not $case_sensitive) {
+                    $c = '_' if $c eq '-';
+                    $c = lc $c;
+                }
+
+                if (not defined $$cursor{$c}) {
+                    $$cursor{$c} = {};
+                    $nodes++;
+                }
+                $cursor = %$cursor{$c};
+            }
+            $$cursor{'val'} = $name;
+        }
+    }
+    #print "\n\n/* $nodes nodes for $chars letters*/\n\n";
+    return %trie;
+}
+
+sub generate_code_from_trie {
+    my $n = shift;
+    my $trieref = shift;
+    my $indent = 1 * $n + 1;
+    my $strcmp = $case_sensitive ? 'strcmp' : 'strcasecmp';
+
+    print "int ossl_param_find_pidx(const char *s)\n{\n" if $n == 0;
+
+    if ($trieref->{'suffix'}) {
+        my $suf = $trieref->{'suffix'};
+
+        printf "%*s if ($strcmp(\"$suf\", s + $n) == 0", $indent, "";
+        if ($case_sensitive) {
+            $suf =~ tr/_/-/;
+            print " || $strcmp(\"$suf\", s + $n) == 0"
```

Comment:
This second strcmp seems odd. I'm not sure why we want this. This seems like a new feature (we don't do this at the moment right?...or do we?)

Reply:
We don't use this currently.  It's used if we're being case insensitive: `_` and `-` matching as identical.